### PR TITLE
Navigation accessibility improvements

### PR DESCRIPTION
**Pull request by jamesjacobs**
This PR includes various navigation accessibility improvements detailed below.

**Main navigation**
- Added `aria-label="primary"`. On entering the nav landmark screenreaders now announce "Main landmark primary navigation" rather than "Main landmark navigation".
- Primary nav links which expand the secondary nav now include the `aria-haspopup`, `aria-expanded` and `aria-controls` attributes. Screen readers now announce that they contain sub menus and are collapsed/expanded (which is updated as they open/close).
- Primary nav links which link to the current page have the `aria-current` which screenreader announce.

**Secondary navigation**
- The close link is now announced as "Close expanded menu", previously they were either not announced or announced as "blank" depending on the screen reader.
- The close link can no longer be tabbed to / gain focus when the menu has been opened and closed. Previously you could tab to the close link after the menu was closed.
- Secondary nav links now have the `aria-describedby` role which appended the secondary nav title to each link when announced.

**Tertiary navigation**
- The close link is now announced as "Close expanded menu", previously it was either not announced or announced as "blank" depending on the screen reader.
- The close link can no longer be tabbed to / gain focus when the menu has been opened and closed. Previously you could tab to the close link after the menu was closed.
- Tertiary navigation links can no longer be tabbed to / gain focus when the menu has been opened and closed. Previously you could tab through all tertiary links even when the menu had been closed.
- Tertiary navigation links which expand the quaternary nav now include the `aria-haspopup`, `aria-expanded` and `aria-controls` attributes. Screen readers now announce that they contain sub menus. They are updated as the quaternary nav opens and closes.
- Tertiary navigation now gets closed when clicking on the main body of content. Previously it did not.

**Quaternary navigation**
- The close link is now announced as "Close expanded menu", previously it was either not announced or announced as "blank" depending on the screen reader.
- Quaternary navigation links can no longer be tabbed to / gain focus when the menu has been opened and closed. Previously you could tab through all tertiary links even when the menu had been closed.
- Quaternary navigation now gets closed when clicking on the main body of content. Previously it did not.

**More button**
- Now has the `aria-expanded` and `aria-controls` attributes. Screen readers now announce that it contains a sub menu and is collapsed/expanded (which is updated as it opens/closes).

**Breadcrumbs**
- Added `aria-label="breadcrumbs". On entering the breadcrumbs nav landmark screenreaders now announce "Breadcrumbs navigation landmark, list with x items " rather than "list with x items".
- Breadcrumb nav links which link to the current page have the `aria-current` which screenreader announce.

**Tabs (inline tab navigation)**
- Added `aria-label="secondary"` and swapped from a `div` to a `nav` element. On entering the tab navigation landmark screenreaders now announce "Secondary navigation landmark, list with x items", rather than "list with x items"
- Tab nav links which link to the current page have the `aria-current` which screenreader announce.

**Mobile menu button**
- Swapped from an `a` to a `button` element
- Now contains `aria-expanded="false"`,  `aria-controls="aria-main-nav"`, `aria-label="Toggle main menu"`. Which is announced as "Button, toggle main menu, contains sub menu" and updates as the menu is opened/closed. Previously it was announced as "link, menu".

**Refactoring 🚜**
I've also refactored a substantial amount of the navigation javascript, mainly to try to simplify and remove some redundant/unneeded or unworking code. I've added tests to cover the changes and also removed some false positive tests + added some missing tests for older code. I resisted re-writing as ES6, due to time.

I also removed the old, unused `tabs` helper.

Tested in:

- [x] JAWS
- [x] NVDA
- [X] Voiceover
- [X] Chrome
- [X] Safari
- [X] Firefox
- [X] IE 11-9

I noticed a display issue in IE9 with the Quaternary navigation, but that was unrelated to my changes so will raise another ticket to address. There are also some tab order improvements that could be made to the navs when sub navs are opened. As the required changes for that would be breaking I've raised them in #787.

### CODE REVIEW

**Comment by jamesjacobs** [comment]
Note to self for when I'm less tired - check JS changes don't throw errors on non-helper generated navs!

**Comment by Stanton** [comment]
# Breaking change quick checklist

The following points should be considered as an early-warning of introducing a breaking change to a Continuum platform product. This is not an exhaustive list of what constitutes a breaking change.

| Does this pull request... | Yes / No |
| ---- | ---- |
| require changes to `main.js` |  |
| require changes to `index.js` |  |
| require changes to `pulsar.scss` |  |
| require changes to `gruntfile.js` |  |
| require changes to `package.json` (not including dev dependencies) |  |
| require changes to `base.html.twig` (or other core views like header/footer) |  |
| require new arguments to be passed to a js component |  |
| require markup changes to maintain functionality |  |

File: js/NavMainComponent.js [hunk 1]
```diff
@@ -18,161 +21,205 @@ NavMainComponent.prototype.init = function () {
...
+            $(this).text('Menu');
+            $(this).attr('aria-expanded', 'false');
+        }
+    });
 
-    component.$primaryNavLinks.on('click', function (e) {
-        var $self = $(this),
-            href = $self.attr('href');
+    // More button is shown when window is too short to display all primary nav items
+    component.$navMain.on('click', '.more-icon > .nav-link', function(event) {
```
Comment (Stanton) [inline js/NavMainComponent.js]: we've used `e` elsewhere in this file, doesn't matter which we use but should be consistent

File: js/NavMainComponent.js [hunk 2]
```diff
@@ -18,161 +21,205 @@ NavMainComponent.prototype.init = function () {
...
-        component.closeNavs();
-    };
+        if ($(this).text() === 'Menu') {
+            $(this).text('Close');
+            $(this).attr('aria-expanded', 'true');
+        } else {
+            $(this).text('Menu');
+            $(this).attr('aria-expanded', 'false');
+        }
+    });
```
Comment (mikedevelops) [inline js/NavMainComponent.js]: micro optimisation here, could we cache `$(this)`, we're creating a lot of object. As far as I know `$(this) === $(this) // false`

File: js/NavMainComponent.js [hunk 3]
```diff
@@ -18,161 +21,205 @@ NavMainComponent.prototype.init = function () {
...
+    component.$window.resize(function () {
+        component.adjustNavItems();
+    });
 
-        component.switchPrimaryNav(href);
-        component.switchSecondaryNav(href);
-        component.$contentMain.on('click', component.closeHandler);
+    // Close navs on main content click
+    component.$contentMain.on('click', function () {
+        var $self = $(this);
```
Comment (Stanton) [inline js/NavMainComponent.js]: any point in caching to `$self` here if it's only used once?

File: js/NavMainComponent.js [hunk 4]
```diff
@@ -18,161 +21,205 @@ NavMainComponent.prototype.init = function () {
...
-    });
+    // Open navigation on mobile
+    component.$mobileMenuButton.on('click', function() {
+        component.$body.toggleClass('open-nav');
+        $(this).toggleClass('open');
 
-    component.closeHandler = function () {
-        component.closeNavs();
-    };
+        if ($(this).text() === 'Menu') {
```
Comment (mikedevelops) [inline js/NavMainComponent.js]: if we _have_ to match on string values, should we normalise them first as a guard? something like `$(this).text().toLowerCase().trim()` or the like

File: js/NavMainComponent.js [hunk 5]
```diff
@@ -18,161 +21,205 @@ NavMainComponent.prototype.init = function () {
...
+    component.$contentMain.on('click', function () {
+        var $self = $(this);
+        component.closeNavs($self);
     });
 
-    component.$secondaryNavLinks.on('click', function () {
-        component.changeActiveSecondaryNavLink($(this).attr('href'));
+    // Open secondary nav on primary nav item click
+    component.$primaryNavLinks.on('click', function (event) {
+        var $self = $(this);
```
Comment (Stanton) [inline js/NavMainComponent.js]: any point in caching to `$self` here if it's only used once?

File: js/NavMainComponent.js [hunk 6]
```diff
@@ -18,161 +21,205 @@ NavMainComponent.prototype.init = function () {
...
+    // Close any previously open navs
+    component.closeSecondaryNav();
+    component.closeTertiaryNav();
+    component.closeQuaternaryNav();
+
+    // If href is a fragment (therefore opens a sub nav), don't add it to the URL because it breaks the back button
+    if (target.indexOf('#') !== -1) {
+        event.preventDefault();
+        
+        // Change aria expanded to true
```
Comment (Stanton) [inline js/NavMainComponent.js]: obvious comment?

File: js/NavMainComponent.js [hunk 7]
```diff
@@ -18,161 +21,205 @@ NavMainComponent.prototype.init = function () {
...
+    var component = this;
 
-NavMainComponent.prototype.closeSubNavs = function () {
+    component.$navTertiary.removeClass('is-open');
+    component.$navTertiary.find('.nav-list').removeClass('is-active');
+
+    // Reset aria-expanded on more button
+    component.$navMain.find('[aria-expanded=true]')
+        .removeClass('is-active')
+        .attr( 'aria-expanded', 'false');
```
Comment (Stanton) [inline js/NavMainComponent.js]: extra space

File: tests/js/web/NavMainComponentTest.js [hunk 8]
```diff
@@ -9,97 +9,88 @@ $.fx.off = !$.fx.off;
...
+            '       <div class="nav-list" data-nav="#three_one">' +
+            '           <ul class="nav-items">' +
+            '               <li class="nav-item">' +
+            '                   <a href="#four_one" class="nav-link">4.1</a>' +
+            '               </li>' +
+            '           </ul>' +
+            '       </div>' +
+            '   </div>' +
+            '</nav>' +
+            '<div class="content-main"></div>'
```
Comment (mikedevelops) [inline tests/js/web/NavMainComponentTest.js]: we should soooo use ES6 string template literals here (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals). It'll mean we can build these DOM structures without all the `+`s

e.g

```
$(`<div>
    <lots>
        <of>
             <dom></dom>
        </of>
    </lots>
</div>`);

File: js/NavMainComponent.js [hunk 9]
```diff
@@ -18,161 +21,205 @@ NavMainComponent.prototype.init = function () {
...
 
-    if (component.$navQuaternary.hasClass('is-open')) {
-        component.$navQuaternary.removeClass('is-open');
-        component.$navQuaternary.find('.nav-list.is-active').removeClass('is-active');
-    } else {
-        component.$navTertiary.removeClass('is-open');
-    }
-};
+    // Reset aria-expanded on tertiary link
+    component.$navTertiary.find('[aria-expanded=true]').attr( 'aria-expanded', 'false');
```
Comment (Stanton) [inline js/NavMainComponent.js]: extra space

**Comment by mikedevelops** [comment]
screen shot 2018-03-29 at 10 00 36

Have we definitely got test coverage for this condition? Looks as though Istanbul says no.

File: js/NavMainComponent.js [hunk 10]
```diff
@@ -283,19 +352,30 @@ NavMainComponent.prototype.lastItemSubstitution = function (numberOfHiddenNavIte
     }
 };
 
-NavMainComponent.prototype.moreIconClickHandler = function () {
+/**
+ * Hide more link if there are no hidden nav items
+ * @param {jQuery} $linkClicked - jQuery object of the more link
```
Comment (Stanton) [inline js/NavMainComponent.js]: should this be `$moreLink`?

File: js/NavMainComponent.js (hunk 1 above)
Comment (jamesjacobs) [reply js/NavMainComponent.js]: changed all to `event`

**Pull request merged**